lib/Dialect/AIG/Transforms/LowerVariadic.cpp

```diff
@@ -0,0 +1,117 @@
+                                     ArrayRef<bool> inverts,
+                                     ConversionPatternRewriter &rewriter) {
+  Value lhs, rhs;
+  switch (operands.size()) {
+  case 0:
+    assert(0 && "cannot be called with empty operand range");
+    break;
+  case 1:
+    if (inverts[0])
+      return rewriter.create<AndInverterOp>(op.getLoc(), operands[0], true);
+    else
+      return operands[0];
+  case 2:
+    lhs = operands[0];
+    rhs = operands[1];
+    return rewriter.create<AndInverterOp>(op.getLoc(), lhs, rhs, inverts[0],
+                                          inverts[1]);
+  default:
+    auto firstHalf = operands.size() / 2;
+    lhs = lowerFullyAssociativeOp(op, operands.take_front(firstHalf),
+                                  inverts.take_front(firstHalf), rewriter);
+    rhs = lowerFullyAssociativeOp(op, operands.drop_front(firstHalf),
+                                  inverts.drop_front(firstHalf), rewriter);
+    return rewriter.create<AndInverterOp>(op.getLoc(), lhs, rhs);
+  }
+}
+
+struct VariadicOpConversion : OpConversionPattern<AndInverterOp> {
+  using OpConversionPattern<AndInverterOp>::OpConversionPattern;
+  using OpAdaptor = typename AndInverterOp::Adaptor;
+  LogicalResult
+  matchAndRewrite(AndInverterOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    if (op.getInputs().size() < 2)
+      return failure();
+
+    // TODO: This is naive implementatino that creates a balanced binary trees.
+    //       WE can improve by analyzing the dataflow and creating a tree that
+    //       improves the critical path or area.
+    auto result = lowerFullyAssociativeOp(op, op.getOperands(),
+                                          op.getInverted(), rewriter);
+    rewriter.replaceOp(op, result);
+    return success();
+  }
+};
+
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// Convert Comb to AIG pass
+//===----------------------------------------------------------------------===//
+
+static void populateLowerVariadicPatterns(RewritePatternSet &patterns) {
+  patterns.add<VariadicOpConversion>(patterns.getContext());
+}
+
+namespace {
+struct LowerVariadicPass : public impl::LowerVariadicBase<LowerVariadicPass> {
+  void runOnOperation() override;
+};
+} // namespace
+
+void LowerVariadicPass::runOnOperation() {
+  ConversionTarget target(getContext());
+  target.addLegalDialect<comb::CombDialect>();
+  target.addIllegalOp<aig::AndInverterOp>();
+  target.addDynamicallyLegalOp<aig::AndInverterOp>(
+      [](AndInverterOp op) { return op.getInputs().size() <= 2; });
+
+  RewritePatternSet patterns(&getContext());
+  populateLowerVariadicPatterns(patterns);
+
+  if (failed(mlir::applyPartialConversion(getOperation(), target,
+                                          std::move(patterns))))
+    return signalPassFailure();
```

Comment:
Could this be a greedy rewriter instead? I don't think you really need the dialect conversion framework. You could check for `op.getInputs().size() <= 2` in the match and return false to indicate that the pattern doesn't apply one an op has been reduced to 2 or fewer operands.